adds properties to dismissible widget

theankurkedia opened the pull request:
Adds 2 properties.
`crossAxisEndOffset` which defines the end offset across the main axis after the card is dismissed.
`rollbackDuration` which defines the duration for card to come back to original position if not dismissed.

Commit:
adds crossAxisEndOffset and rollbackDuration properties to dismissible widget

theankurkedia committed:
re-added removed license comment

theankurkedia committed:
corrected license comments

Hixie commented on packages/flutter/lib/src/widgets/dismissible.dart:
```diff
@@ -1,7 +1,8 @@
+// found in the LICENSE file.		
 
+ 
 import 'package:flutter/foundation.dart';
```
looks like you have some accidental whitespace changes near the top (new spaces and new lines)

Hixie commented on packages/flutter/lib/src/widgets/dismissible.dart:
```diff
@@ -134,6 +136,13 @@ class Dismissible extends StatefulWidget {
+
+  /// Defines the end offset across the main axis after the card is dismissed.
+  /// If non-zero value is given then widget moves in cross direction depending on whether it is positive or negative.
+  final double crossAxisEndOffset;
```
Add a blank line in the documentation after the first sentence (the first paragraph is used when we need a short description).

Hixie commented:
This looks good. One thing it will need is tests, so that we don't accidentally break this or remove the feature at some point. Anything you care about making sure we don't break, please add a test to verify that it works. You can add the tests to: https://github.com/flutter/flutter/blob/master/packages/flutter/test/widgets/dismissible_test.dart

Commit:
Adds test for rollbackDuration and crossAxisEndOffset

Commit:
added tests and comments

theankurkedia commented:
@Hixie I have added the tests. Both the properties are being covered in the code.

Hixie commented on packages/flutter/test/widgets/dismissible_test.dart:
```diff
@@ -143,6 +147,58 @@ Future<Null> dismissItem(WidgetTester tester, int item, {
+      downLocation = tester.getTopLeft(finder) + const Offset(0.0, 0.1);
+      upLocation = tester.getTopLeft(finder) + const Offset(0.0, 0.2);
+      break;
+    default:
```
you can remove the default; the analyzer will then warn us at analyze time if there's an unsupported gesture direction

Hixie commented on packages/flutter/test/widgets/dismissible_test.dart:
```diff
@@ -143,6 +147,58 @@ Future<Null> dismissItem(WidgetTester tester, int item, {
+    case AxisDirection.left:
+      // getTopRight() returns a point that's just beyond itemWidget's right
+      // edge and outside the Dismissible event listener's bounds. Just nudge
+      // the itemWidget before it's threshold.
```
its

Hixie commented on packages/flutter/lib/src/widgets/dismissible.dart:
```diff
@@ -83,6 +82,8 @@ class Dismissible extends StatefulWidget {
     this.direction: DismissDirection.horizontal,
     this.resizeDuration: const Duration(milliseconds: 300),
     this.dismissThresholds: const <DismissDirection, double>{},
+    this.rollbackDuration: const Duration(milliseconds: 300),
```
this should be 200ms, right? To match the previous _kDismissDuration?

theankurkedia replied:
Yes, you are right.

Hixie commented on packages/flutter/test/widgets/dismissible_test.dart:
```diff
@@ -558,4 +614,29 @@ void main() {
+    expect(find.text('1'), findsOneWidget);
+    expect(dismissedItems, isEmpty);
+  });
+}
```
maybe i misunderstand the test but it doesn't seem to me that it verifies that things actually go diagonally or that they actually take rollbackDuration time to go away. For example, if someone replaced `widget.crossAxisEndOffset` with `0.0` in `_updateMoveAnimation`, would the test fail? Similarly, would your test fail if the duration was not precisely honoured, but instead e.g. the rollback happened in half the specified time? or double the specified time?

theankurkedia replied:
You are right again, what all it does is check if the widget is still there after giving it a small nudge. For `rollbackDuration`, I am trying to find a more specific way to test the features but I am caught up in it. What I have in mind now is to check the duration once the widget is released from drag, but I can't seem to find a way to do it. And for `crossAxisEndOffset`, I am trying to get the `topLeft` or `topRight` offset and check, but I am not sure if it is the right way to go ahead. I haven't had success doing it yet.

Hixie replied:
For the duration, you can just pump a duration less than the rollback, verify it hasn't finished rolling back, then pump a duration that takes you past the end of the rollback, and verify that it _has_ finished.

For the positions, yet, you can use tester.getTopLeft and so forth to find coordinates.

Commit:
modified tests for dismissible widgets

theankurkedia commented:
First, I have changed the name of `rollbackDuration` with `movementDuration` because I observed that while using the widget, this duration was also applicable when the fling is given and widget is dismissed.

theankurkedia commented:
Second, as you suggested, I have modified tests. I have added 1 case here, in which the widget rolls back to original position (i.e. `movementDuration`). When we give I small nudge to it, the widget comes back to original place, but never goes out of `finder`, so the widget is always present. What I added was that while we give a fling to widget, it goes out of window after `movementDuration` and starts resizing. I have added checks whether we have a widget before completion of duration and after that duration and it is working fine. Although, I am still not that confident about this because the tests are still passing even when I change the default movementDuration.

Hixie commented:
Thanks for your contribution!

If you like, you can submit additional tests in the future, to help ensure we don't accidentally regress part of your patch. (Anything that's not tested will eventually break, that's just a rule of software development.)
